src/server/json_family.cc

```diff
@@ -157,6 +159,98 @@ OpResult<json> GetJson(const OpArgs& op_args, string_view key) {
   return decoder.get_result();
 }
 
+string ConvertToJsonPointer(string_view json_path) {
+  if (json_path.empty() || json_path[0] != '$') {
+    return "";
+  }
+
+  // except the supplied string is compatible with JSONPath syntax.
+  // Each item in the string is a left bracket followed by
+  // numeric or '<key>' and then a right bracket.
+  vector<string> parts;
+  bool invalid_syntax = false;
+  std::size_t i = 1;
+  while (i < json_path.size()) {
+    bool is_array = false;
+    bool is_object = false;
+
+    // check string size is sufficient enough for at least one item.
+    if (i + 2 >= json_path.size()) {
+      invalid_syntax = true;
+      break;
+    }
+
+    if (json_path[i] == '[') {
+      if (json_path[i + 1] == '\'') {
+        is_object = true;
+        i += 2;
+      } else if (isdigit(json_path[i + 1])) {
+        is_array = true;
+        i += 1;
+      } else {
+        invalid_syntax = true;
+        break;
+      }
+    } else {
+      invalid_syntax = true;
+      break;
+    }
+
+    if (is_array) {
+      auto end_pos = json_path.find(']', i);
+      if (end_pos == string::npos) {
+        invalid_syntax = true;
+        break;
+      }
+
+      parts.emplace_back(json_path.substr(i, end_pos - i));
+      i = end_pos + 1;
+
+    } else if (is_object) {
+      // find the index of the right bracket.
+      auto end_val_idx = string::npos;
+      auto current_idx = i;
+      while (end_val_idx == string::npos) {
+        if (current_idx + 1 >= json_path.size()) {
+          break;
+        }
+
+        // ignore escaped character (e.g. \']).
+        if (json_path[current_idx] == '\\') {
+          current_idx += 2;
+        }
+
+        else if (json_path[current_idx] == '\'' && json_path[current_idx + 1] == ']') {
+          end_val_idx = current_idx++;
+        }
+
+        else {
+          current_idx++;
+        }
+      }
+
+      if (end_val_idx == string::npos) {
+        invalid_syntax = true;
+        break;
+      }
+
+      parts.emplace_back(json_path.substr(i, end_val_idx - i));
```

Comment:
it seems that it's possible to refactor the code in the loop into a very nice function:
`string_view GetNextIndex(string_view str)` that assumes that it gets a string right after `[` and returns 
1. an empty string_view in case of invalid syntax
2. an index, i.e. the string part until the closing `]`. 

This will make the code in this function much clearer. In addition, it will remove the repeated
```
 invalid_syntax = true;
 break;
``` 
blocks because you can just return `string_view{}` from the inner function instead.